### PR TITLE
ceph-volume,python-common: Data allocate fraction

### PR DESCRIPTION
**Pull request by PepperJo**
This PR allows to partially allocate a data device when preparing/deploying an OSD with ceph-volume or through the device group specification. The fraction can be between (0,1.0] where a fraction of e.g. 0.8 will allocate 80% of the data device, i.e. the lvm size will only account for 80% of the device's capacity. The rest of the device will not be allocated. This can be used to alleviate write performance degradation of Flash SSDs.
This PR adds `--data-allocate-fraciton` to ceph-volume lvm batch and the `data_allocate_fraction` feature to the device group specification.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
Added [needs-review]

File: src/ceph-volume/ceph_volume/devices/lvm/batch.py [hunk 1]
```diff
@@ -269,6 +270,17 @@ def __init__(self, argv):
                   ' if more slots then osds-per-device are specified, slots'
                   'will stay unoccupied'),
         )
+        def data_allocate_fraction(pct):
```
Comment (andrewschoen) [inline src/ceph-volume/ceph_volume/devices/lvm/batch.py:273]: Could you move this function to `utils.arg_validators` please? You can see an example usage with `arg_validators.ValidDevice` here: https://github.com/ceph/ceph/pull/40659/files#diff-37d41dc2d64315970e0a32c6d84be634f82c62a15c4e3144020ea2be88641601R200

**Review by andrewschoen — CHANGES_REQUESTED** [review]
Overall this looks good. I think we need to add tests for `data_allocate_fraction` though. Tests for other argument validators can be found in `tests.util.test_arg_validators` as an example.

**Comment by andrewschoen** [comment]
jenkins test ceph-volume all

**Comment by andrewschoen** [comment]
jenkins test ceph-volume tox

**Comment by andrewschoen** [comment]
jenkins test ceph-volume all

**Comment by andrewschoen** [comment]
jenkins test ceph-volume tox

**Comment by andrewschoen** [comment]
jenkins test ceph-volume tox

**Comment by PepperJo** [comment]
@andrewschoen thanks for your review. I moved the function as suggested and added a few tests.

**Comment by andrewschoen** [comment]
jenkins test ceph-volume all

**Comment by andrewschoen** [comment]
> @andrewschoen thanks for your review. I moved the function as suggested and added a few tests.

@PepperJo thank you for the work! These changes look great. If the CI finishes clean, I'll approve this.

**Comment by PepperJo** [comment]
@andrewschoen I checked the failed tests and if I interpret them correctly there seems to be a problem with `fetch ceph red hat development repository`? That seems to be unrelated to my changes.

**Comment by andrewschoen** [comment]
jenkins test ceph-volume all

**Comment by andrewschoen** [comment]
> @andrewschoen I checked the failed tests and if I interpret them correctly there seems to be a problem with `fetch ceph red hat development repository`? That seems to be unrelated to my changes.

@PepperJo I agree, they do look unrelated. It looks like there was some trouble accessing some dev repos to test with, which I suspect is transient. I'll try again as these tests did all pass on previous runs.

**Comment by andrewschoen** [comment]
jenkins test ceph-volume lvm centos8-filestore-dmcrypt

**Comment by andrewschoen** [comment]
jenkins test ceph-volume lvm centos8-bluestore-dmcrypt

**Comment by andrewschoen** [comment]
jenkins test ceph-volume lvm centos8-bluestore-dmcrypt

**Review by andrewschoen — APPROVED** [review]

**Comment by PepperJo** [comment]
@andrewschoen Thank you!

**Pull request merged**